src/config/file/v1.rs

```diff
@@ -64,20 +66,41 @@ impl RawConfigV1 {
     }
 }
 
-fn validate(config: RawConfigV1) -> Result<super::Config, Report<Error>> {
+async fn validate(config: RawConfigV1) -> Result<super::Config, Report<Error>> {
     let endpoint = fossa::Endpoint::try_from(config.endpoint).change_context(Error::Validate)?;
     let key = fossa::Key::try_from(config.integration_key).change_context(Error::Validate)?;
     let api = fossa::Config::new(endpoint, key);
     let debugging = debug::Config::try_from(config.debugging).change_context(Error::Validate)?;
+
+    // Converts and validates the parsed integrations
+    // The first map tries to convert each integration into an integration object after validation
+    // The second map does a best effort approach to inject main/master branch into watched_branches if watched_branches is empty and import_branches is set to true
     let integrations = config
         .integrations
         .into_iter()
         .map(remote::Integration::try_from)
+        .map(|res| async {
+            match res {
+                Ok(integration) => {
+                    Ok(remote::Integration::set_watched_branches(&integration).await)
+                }
+                Err(report) => Err(report),
+            }
```

Comment:
This pattern of "If the value is Ok, do something to the value inside it otherwise return it unchanged" is common enough that there's a method to help with it called [map](https://doc.rust-lang.org/std/result/enum.Result.html#method.map). 

I think its use here would be something like this:
```rust
res.map(|integration| remote::Integration::set_watched_branches(integration).await)
```

Rust doesn't use the same terminology, but this roughly corresponds to `fmap` that we've been working with in Haskell. Compare with the [`Either`](https://hackage.haskell.org/package/base-4.18.1.0/docs/Prelude.html#t:Either) type, which is roughly equivalent to Rust's `Result` type. 

AFAIK rust doesn't have a general interface for things that are mappable the way that Haskell does (this is what `Functor` is) but it often implements functions like `map` that allow it for a specific type.

Reply:
There's a few issues with this in general:
- You have synchronous errors, but you're putting them in a future. Just handle them up front.
- You are creating a nested `Result<Result<_, _>, _>` structure here. Handling the synchronous errors would allow you to `flat_map` and avoid the nested structure.

For example, it'd look like this:
```rust
let integrations =
        config
            .integrations
            .into_iter()
            .map(remote::Integration::try_from)
            .collect::<Result<Vec<_>, _>>()
            .change_context(Error::Validate)?
            .into_iter()
            .map(|integration| async move {
                remote::Integration::set_watched_branches(&integration).await
            });
```

But note that this is primarily due to the fact that you really have a new function that _should be called in your constructor_. It's awkward because it's secretly a constructor responsibility, but you're not putting it there.